File: js/taceo-proof-client/src/WebSocket.ts
/**
 * The stop strategy when waiting for the job result.
 */
export type StopStrategy =
  /** Stop if one node responded with a result. */
  | "First"
  /** Stop if the majority of nodes responded with a result. */
  | "Majority"
  /** Stop if all nodes responded with a result. */
  | "All";

/**
 * The WebSocket subscribe request to wait for a job result.
 *
 * This request is sent to the WebSocket server to subscribe to updates
 * for a specific job execution.
 */
interface SubscribeExecutionRequest {
  /** The unique identifier of the job execution. */
  execution_id: string;

  /** The stop strategy to determine when to stop waiting for job results. */
  stop_on_finished_reports: StopStrategy;

  /** Whether to include status updates in the WebSocket messages. */
  with_status_updates?: boolean;
}

/**
 * The status of a job.
 */
type JobStatus =
  | "Pending"
  | "Running"
  | "Finished"
  | "Cancelled";

/**
 * The result of a job, including signatures of the nodes that handled the job.
 */
export interface SignedResults {
  /** The signatures of each node used to compute the job (identified by their id). */
  signatures: { [key: number]: string };

  /** The proof as JSON (for CircomGroth16 proofs) or base64 encoded ark_serialized `ark_groth16::Proof` (for LibsnarkGroth16 proofs). */
  proof: string;

  /** The array of public inputs as JSON (for CircomGroth16 proofs) or base64 encoded ark_serialized `Vec<P::ScalarField>` (for LibsnarkGroth16 proofs). */
  public_inputs: string;
}

/**
 * An error result of a job.
 *
 * Provides details about a failure encountered during the execution of a job.
 */
interface FailedReason {
  /** The node provider that encountered the error. */
  node_provider: number;
  /** The error string describing the failure. */
  error: string;
  /** The signature of the error and the node provider. */
  signature: string;
}

/**
 * The different messages received over the WebSocket connection.
 */
type WebSocketMessage =
  /** The job completed successfully with signed results. */
  | { Success: SignedResults }
  /** A status update for the job. */
  | { Update: JobStatus }
  /** The job failed with a reason. */
  | { Failed: FailedReason }
  /** The job was cancelled. */
  | { Cancelled: null }
  /** An error occurred during the WebSocket communication. */
  | { Err: string };

/**
 * Fetches the result of a job execution using a WebSocket connection.
 *
 * This function connects to the specified WebSocket `url` and subscribes to updates
 * for the job identified by `jobId`. It listens for messages from the server and
 * processes them based on the provided `stopStrategy`. The function returns the
 * signed results of the job execution if successful, or an error if the job fails
 * or the connection is closed unexpectedly.
 *
 * @param url - The WebSocket URL to connect to for job updates.
 * @param jobId - The unique identifier of the job whose results are being fetched.
 * @param stopStrategy - The strategy to determine when to stop waiting for job results.
 *
 * @returns A promise that resolves to a `SignedResults` object containing the proof,
 * public inputs, and signatures from the nodes that handled the job.
 *
 * @throws This function throws an error if:
 * - The WebSocket connection fails to establish.
 * - The server sends invalid or unexpected data.
 * - The job fails, is cancelled, or encounters an error on a node.
 * - The server closes the WebSocket connection unexpectedly.
 */
export async function fetchJobResult(
  url: string,
  jobId: string,
  stopStrategy?: StopStrategy,
): Promise<SignedResults> {
  return new Promise((resolve, reject) => {
    const socket = new WebSocket(url);
    socket.onopen = () => {
      const request: SubscribeExecutionRequest = {
        execution_id: jobId,
        stop_on_finished_reports: stopStrategy ? stopStrategy! : "First",
        with_status_updates: false,
      };
      socket.send(JSON.stringify(request));
    };
    socket.onmessage = (event) => {
      const msg: WebSocketMessage = JSON.parse(event.data);
      if ("Success" in msg) {
        resolve(msg.Success);
      } else if ("Failed" in msg) {
        reject(new Error(msg.Failed.error));
      } else if ("Cancelled" in msg) {
        reject(new Error("Job was cancelled!"));
      } else if ("Err" in msg) {
        reject(new Error(msg.Err));
      }
      socket.close();
    };
    socket.onerror = (event) => {
      reject(new Error(`WebSocket error: ${event}`));
      socket.close();
    };
  });
}

File: rust/taceo-proof-client/src/websocket.rs
use std::fmt;

use eyre::Context as _;
use futures::{SinkExt as _, StreamExt as _};
use intmap::IntMap;
use serde::{Deserialize, Serialize};
use tokio_tungstenite::tungstenite::Message;
use uuid::Uuid;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
/// The status of a job.
pub enum JobStatus {
    Pending,
    Running,
    Finished,
    Cancelled,
}

#[derive(Serialize, Deserialize)]
/// The result of a job, including signatures of the nodes that handled the job.
pub(crate) struct SignedResults {
    /// The signatures of each node used to compute the job (identified by their id).
    pub signatures: IntMap<i32, String>,
    /// The proof as JSON (for CircomGroth16 proofs) or base64 encoded ark_serialized `ark_groth16::Proof` (for LibsnarkGroth16 proofs).
    pub proof: String,
    /// The array of public inputs as JSON (for CircomGroth16 proofs) pr base64 encoded ark_serialized `Vec<P::ScalarField>` (for LibsnarkGroth16 proofs).
    pub public_inputs: String,
}

#[derive(Serialize, Deserialize, Debug)]
/// A error result of a job.
struct FailedReason {
    /// The node provider that encountered the error.
    pub node_provider: i32,
    /// The error string.
    pub error: String,
    /// The signature of the error and the node provider.
    pub signature: String,
}

impl fmt::Debug for SignedResults {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let node_providers = self
            .signatures
            .keys()
            .map(|k| k.to_string())
            .collect::<Vec<_>>();
        f.write_fmt(format_args!("Result from: {node_providers:?}"))
    }
}

#[derive(Serialize, Deserialize, Debug)]
/// The different messages we receive over the ws connection.
enum WebSocketMessage {
    Success(SignedResults),
    Update(JobStatus),
    Failed(FailedReason),
    Cancelled,
    Err(String),
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, Default)]
/// The stop strategy when waiting for the job result.
pub enum StopStrategy {
    #[default]
    /// Stop if one node responded with a result.
    First,
    /// Stop if the majority of nodes responded with a result.
    Majority,
    /// Stop if the all nodes responded with a result.
    All,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
/// The ws subscribe request to wait for a job result.
struct SubscribeExecutionRequest {
    pub execution_id: Uuid,
    pub stop_on_finished_reports: StopStrategy,
    pub with_status_updates: Option<bool>,
}

pub(crate) async fn fetch_job_result(
    url: &str,
    job_id: Uuid,
    stop_strategy: StopStrategy,
) -> eyre::Result<SignedResults> {
    // subscribe with ws to get job updates
    let (mut ws_stream, _) = tokio_tungstenite::connect_async(url)
        .await
        .context("while connecting to endpoint")?;
    ws_stream
        .send(Message::text(
            serde_json::to_string(&SubscribeExecutionRequest {
                execution_id: job_id,
                stop_on_finished_reports: stop_strategy,
                with_status_updates: Some(false),
            })
            .expect("can serialize"),
        ))
        .await
        .context("while sending subscribe request")?;
    if let Some(msg) = ws_stream.next().await {
        match msg {
            Ok(Message::Text(text)) => {
                let msg: WebSocketMessage =
                    serde_json::from_str(&text).context("invalid data from server")?;
                match msg {
                    WebSocketMessage::Success(signed_results) => Ok(signed_results),
                    WebSocketMessage::Failed(FailedReason {
                        node_provider,
                        error,
                        signature: _,
                    }) => eyre::bail!("node {node_provider} error: {error:?}"),
                    WebSocketMessage::Cancelled => eyre::bail!("job was cancelled"),
                    WebSocketMessage::Err(err) => eyre::bail!(err),
                    WebSocketMessage::Update(_) => eyre::bail!("unexpected update"),
                }
            }
            Ok(Message::Close(_)) => {
                eyre::bail!("server closed stream");
            }
            Ok(_) => {
                eyre::bail!("server sent invalid data");
            }
            Err(err) => Err(err.into()),
        }
    } else {
        eyre::bail!("server closed stream");
    }
}
